compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAccessToAcquireRelease.cpp

```diff
@@ -0,0 +1,248 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Iterators.h"
+
+#define DEBUG_TYPE "iree-amdaie-access-to-acquire-release"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Walk all read access operations within the core operations and insert
+/// semaphore acquire and release stubs. Acquire operations will be inserted
+/// at the location of the access operation and release operations will be
+/// inserted before the next access or at the end of the block.
+LogicalResult readAccessToAcquireRelease(Operation *parentOp) {
+  IRRewriter rewriter(parentOp->getContext());
+
+  SmallVector<AMDAIE::CoreOp> coreOps;
+  parentOp->walk([&](AMDAIE::CoreOp coreOp) { coreOps.push_back(coreOp); });
+
+  // Map from DMA source/target logical objectFifos to those respective DMA
+  // operations.
+  DenseMap<Value, AMDAIE::CircularDmaCpyNdOp> logicalObjectFifoToDma;
+  parentOp->walk([&](AMDAIE::CircularDmaCpyNdOp dmaOp) {
+    logicalObjectFifoToDma[dmaOp.getSource()] = dmaOp;
+    logicalObjectFifoToDma[dmaOp.getTarget()] = dmaOp;
+  });
+
+  for (AMDAIE::CoreOp coreOp : coreOps) {
```

Comment:
Why not directly walk the coreOps?

Reply:
It's to avoid nested walks both returning `WalkResult` for error propagation. I think the code is more clear like this.